North-Seattle-College/ad440-winter2021-thursday-repo
# NSG Deploy Script

### Description
Resolves  #15 - Create NSG script

**This PR introduces a script that can be run to create a Network Security Group in Azure**. 

The script also contains a basic region resource check, because I've experienced issues when working in close to over allocated regions in Azure. This means if your initially chosen region has a lot of existing resources, the script will prompt you if you would like to choose another region. However it's a work in progress, and will only allow you to check one other region's resources before making you choose a region for deployment.

**Note:** that settings for the NSG (tags, security rules...) can be altered in the parameters.json before running the script.

### To Test:
**This script requires powershell, to install check out: [Powershell Install](https://docs.microsoft.com/en-us/powershell/scripting/install/installing-powershell?view=powershell-7.1 )**
1. While in Powershell, in the directory of the script, enter $./deploy-nsg.ps1
2. You will be prompted to enter your service principal, resource group and deployment fields
   - Username (found in Azure Key Vault as appid)
   - Tenant ID
   - Resource group name that you will create/use
   - Region
   - Deployment Name
   - Template.json file path
   - Parameter.json file path
3. After you have entered the following fields you will be asked for your credentials password (found in Azure Key Vault as appsecret)
4. Depending on if your chosen region is close to its resource limit you may or may not see the next prompt asking you if you still want to choose this region for your resources. (You can test the limit checker by changing the ResourceLimit param to a lower value)
5. If not then the script should hopefully create your NSG given everything works and the user inputs were correct!
6. Check that your resource group and NSG were properly created in Azure.
7. **Last but not least be sure to delete the resources created by this script after you are done!**

(Time table format taken from Kevin P's PR)
|**DATE**|**ACTIVITY**|**TIME**|
|---|:---|:---|
|1/10|Watch lecture video for insight on steps|1 hours|
|1/14|Install required extensions and updates for Azure/powershell|2 hours|
|1/14|Code the script/parameters/template and submit PR|5 hour|

Total time: ~8hr
